aten/src/ATen/native/cuda/ReduceOpsKernel.cu

```diff
@@ -470,6 +470,203 @@ void cummin_helper_cuda(const Tensor& self, Tensor& values, Tensor& indices, int
+__host__ void scan_outer_dim(const Tensor& self, Tensor& result,
+                                       int dim, scalar_t init, BinaryFunction binary_op) {
+  int row_size = self.size(dim);
+  auto sizes = self.sizes();
+
+  // Treat all outer dimensions (i.e. dim_ < dim) as one.
+  int num_orows = std::accumulate(sizes.begin(), sizes.begin() + dim, 1, std::multiplies<int>());
+
+  // Treat all inner dimensions (i.e. dim > dimension) as one.
+  int num_irows = std::accumulate(sizes.begin() + dim + 1, sizes.end(), 1, std::multiplies<int>());
+
+  dim3 threads(std::min(512, int(num_irows)));
+  int maxGridDim = at::cuda::getCurrentDeviceProperties()->maxGridSize[0];
+  dim3 grid(std::min(maxGridDim, num_orows), std::min(maxGridDim, cuda::ATenCeilDiv(num_irows, int(threads.x))));
+
+  tensor_kernel_scan_outer_dim<scalar_t><<<grid, threads, 0, at::cuda::getCurrentCUDAStream()>>>(
+    result.data_ptr<scalar_t>(), self.data_ptr<scalar_t>(),
+    num_orows, num_irows, row_size, init, binary_op);
+  AT_CUDA_CHECK(cudaGetLastError());
+}
+
+template <typename scalar_t, class BinaryFunction>
+void scan_innermost_dim(const Tensor& self, Tensor& result, scalar_t init, BinaryFunction binary_op) {
+  int ndim = self.dim();
+  // Treat all outer dimensions as a single dimension.
+  int row_size = self.size(ndim - 1);
+  int num_rows = self.numel() / row_size;
+
+  dim3 threads(16, 32);
+  dim3 grid(std::min(at::cuda::getCurrentDeviceProperties()->maxGridSize[0], cuda::ATenCeilDiv(num_rows, int(threads.y))));
+
+  tensor_kernel_scan_innermost_dim<scalar_t, 16, 32><<<grid, threads, 0, at::cuda::getCurrentCUDAStream()>>>(
+    result.data_ptr<scalar_t>(), self.data_ptr<scalar_t>(),
+    num_rows, row_size, init, binary_op);
+  AT_CUDA_CHECK(cudaGetLastError());
+}
+
+template<typename scalar_t, typename BinaryFunction>
+void scan_dim(const Tensor& self, Tensor& result,
+     int64_t dim, scalar_t init, BinaryFunction binary_op) {
+  int ndim = self.dim();
+  Tensor self_ = self.contiguous();
+  result = result.contiguous();
+  if (dim == ndim - 1) {
+    scan_innermost_dim<scalar_t>(self_, result, init, binary_op);
+  } else {
+    scan_outer_dim<scalar_t>(self_, result, dim, init, binary_op);
+  }
+}
+
+Tensor& _logcumsumexp_out_cuda(Tensor& result, const Tensor& self, int64_t dim) {
+  result.resize_(self.sizes());
+  if (self.dim() == 0) {
+    result.fill_(self);
+    return result;
+  }
+  auto wrap_dim = maybe_wrap_dim(dim, self.dim());
+
+  TensorArg output_arg{ result, "output", 1 };
+  TensorArg input_arg{ self, "input", 2 };
+  checkAllSameGPU("logcumsumexp", {output_arg, input_arg});
+
+  AT_DISPATCH_FLOATING_TYPES_AND(at::ScalarType::Half,
+    self.scalar_type(), "logcumsumexp_cuda", [&]() {
+    scalar_t init = -std::numeric_limits<scalar_t>::infinity();
+    auto log_add_exp = [] GPU_LAMBDA (scalar_t x, scalar_t y) -> scalar_t {
+      return std::log(1 + std::exp(std::min(x, y) - std::max(x, y))) +
```

Comment:
Minor point: maybe you should use the log1p function (log (1 + x)) instead of log.

Reply:
Makes sense. Thanks.

Resolved.